File: templates/java/04-DomainToolSupportAgent.java
package com.example.embabel.starter;

import com.embabel.agent.api.annotation.AchievesGoal;
import com.embabel.agent.api.annotation.Action;
import com.embabel.agent.api.annotation.Agent;
import com.embabel.agent.api.common.OperationContext;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import org.springframework.ai.tool.annotation.Tool;

/**
 * Starter: domain object as LLM tool (DICE).
 */
@Agent(description = "TODO: support agent using domain tools")
public class DomainToolSupportAgent {

    public record Customer(Long id, String name, float balance, float pending) {
        @Tool(description = "Get balance; includePending=true adds pending amounts")
        public float balance(boolean includePending) {
            return includePending ? balance + pending : balance;
        }

        /** Not a tool — stays app-private. */
        public String internalNote() {
            return pending > balance ? "HIGH_PENDING" : "OK";
        }
    }

    public record SupportInput(
            @JsonPropertyDescription("Customer id") Long customerId,
            @JsonPropertyDescription("Customer question") String query
    ) {
    }

    public record SupportOutput(String advice, boolean escalate, int risk) {
    }

    /** TODO: replace with real repository. */
    private Customer find(Long id) {
        return new Customer(id, "Ada", 1000f, 50f);
    }

    @AchievesGoal(description = "Customer supported")
    @Action
    public SupportOutput support(SupportInput input, OperationContext context) {
        var customer = find(input.customerId());
        var note = customer.internalNote();
        return context.ai()
                .withDefaultLlm()
                .withToolObject(customer)
                .createObject(
                        """
                                Advise %s. Use balance tools when needed.
                                Internal note: %s
                                Query: %s
                                """.formatted(customer.name(), note, input.query()),
                        SupportOutput.class
                );
    }
}

File: templates/kotlin/04-DomainToolSupportAgent.kt
package com.example.embabel.starter

import com.embabel.agent.api.annotation.AchievesGoal
import com.embabel.agent.api.annotation.Action
import com.embabel.agent.api.annotation.Agent
import com.embabel.agent.api.common.OperationContext
import com.embabel.agent.api.common.createObject
import com.fasterxml.jackson.annotation.JsonPropertyDescription
import org.springframework.ai.tool.annotation.Tool

/**
 * Starter: domain object as LLM tool (DICE).
 */
@Agent(description = "TODO: support agent using domain tools")
class DomainToolSupportAgent {

    data class Customer(
        val id: Long,
        val name: String,
        val balance: Float,
        val pending: Float,
    ) {
        @Tool(description = "Get balance; includePending=true adds pending amounts")
        fun balance(includePending: Boolean): Float =
            if (includePending) balance + pending else balance

        /** Not a tool — stays app-private. */
        fun internalNote(): String =
            if (pending > balance) "HIGH_PENDING" else "OK"
    }

    data class SupportInput(
        @param:JsonPropertyDescription("Customer id")
        val customerId: Long,
        @param:JsonPropertyDescription("Customer question")
        val query: String,
    )

    data class SupportOutput(val advice: String, val escalate: Boolean, val risk: Int)

    /** TODO: replace with real repository. */
    private fun find(id: Long): Customer =
        Customer(id, "Ada", 1000f, 50f)

    @AchievesGoal(description = "Customer supported")
    @Action
    fun support(input: SupportInput, context: OperationContext): SupportOutput {
        val customer = find(input.customerId)
        val note = customer.internalNote()
        return context.ai()
            .withDefaultLlm()
            .withToolObject(customer)
            .createObject(
                """
                Advise ${customer.name}. Use balance tools when needed.
                Internal note: $note
                Query: ${input.query}
                """.trimIndent(),
            )
    }
}
